xls/modules/rle/rle_enc_adv.x

```diff
@@ -0,0 +1,485 @@
+// Copyright 2023 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a parametric multisymbol RLE encoder
+//
+// The encoder uses Run Length Encoding (RLE) to compress the input stream of
+// repeating symbols to the output stream that contains the symbols and
+// the number of its consecutive occurrences in the input stream.
+// Both the input and the output channels use additional `last` flag
+// that indicates whether the packet ends the transmission. After sending
+// the last packet encoder dumps all the data to the output stream.
+// The behavior of the encoder is presented on the waveform below:
+
+// This encoder is implemented as a net of 4 proc.
+// 1. Reduce proc - this process takes incoming symbols and symbol_valid
+// and reduces them into symbol count pairs. This step is stateless.
+// 2. Realign proc - this process moves pairs emitted from the reduce step
+// so that they are aligned to the left, it also calculates propagation
+// distance for the first pair.
```

Comment:
can you also explain what "propagation distance" mean here? (it's not clear, at least to me, from just reading the example)

Reply:
Here propagation distance is the number of consecutive pairs, starting form first pair, that have the same symbol value as the first one and have their counter maxed out. This way 'Core proc' can only check "propagation distance"th pair value and counter. 
* If value differs from the internal state, then all pairs from [0:"propagation distance") will differ and core should combine payload with its state.
* If value is equal internal state, then  "Core proc" only checks if internal state counter + "propagation distance"th pair counter overflow.